amago/envs/builtin/xland_minigrid.py

```diff
@@ -0,0 +1,297 @@
+        self.x_sample = jax.jit(
+            jax.vmap(self.benchmark.sample_ruleset, in_axes=(0,)),
+            device=self.jax_device,
+        )
+        self.x_reset = jax.jit(
+            jax.vmap(self.x_env.reset, in_axes=(0, 0)), device=self.jax_device
+        )
+        self.x_step = jax.jit(
+            jax.vmap(self.x_env.step, in_axes=(0, 0, 0)), device=self.jax_device
+        )
+        self.x_swap_rules = jax.jit(_swap_rules, device=self.jax_device)
+        self.x_swap_tsteps = jax.jit(_swap_trees, device=self.jax_device)
+
+        obs_shapes = self.x_env.observation_shape(self.env_params)
+        self.max_steps_per_episode = self.env_params.max_steps
+        self.action_space = gym.spaces.Discrete(self.x_env.num_actions(self.env_params))
+        self.observation_space = gym.spaces.Dict(
+            {
+                "grid": gym.spaces.Box(
+                    low=0, high=14, shape=obs_shapes["img"], dtype=np.uint8
+                ),
+                "direction_done": gym.spaces.Box(
+                    low=0, high=1, shape=(5,), dtype=np.uint8
+                ),
+                "goal": gym.spaces.Box(
+                    low=0, high=14, shape=obs_shapes["goal_encoding"], dtype=np.uint8
+                ),
+            }
+        )
+        # establish placeholder values for resets
+        ruleset = self.x_sample(self.new_ruleset_keys())
+        self.env_params = self.env_params.replace(ruleset=ruleset)
+        self.x_timestep = self.x_reset(self.env_params, self.new_reset_keys())
+        self.current_episode = jnp.zeros(
+            self.parallel_envs, dtype=jnp.int32, device=self.jax_device
+        )
+        self.episode_return = jnp.zeros(
+            self.parallel_envs, dtype=jnp.float32, device=self.jax_device
+        )
+        self.episode_steps = jnp.zeros(
+            self.parallel_envs, dtype=jnp.int32, device=self.jax_device
+        )
+
+    def new_reset_keys(self):
+        keys = dp(
+            jax.random.split(self._reset_key, num=self.parallel_envs + 1),
+            self.jax_device,
+        )
+        self._reset_key = keys[0]
+        return keys[1:]
+
+    def new_ruleset_keys(self):
+        keys = dp(
+            jax.random.split(self._ruleset_key, num=self.parallel_envs + 1),
+            self.jax_device,
+        )
+        self._ruleset_key = keys[0]
+        return keys[1:]
+
+    @property
+    def suggested_max_seq_len(self):
+        return (self.max_steps_per_episode + 1) * self.k_shots
+
+    def get_obs(self):
+        obs = self.x_timestep.observation
+        done = self.x_timestep.last()
+        direction_done = np.concatenate(
+            (obs["direction"], done[:, np.newaxis]), axis=-1
+        )
+        # it seems more in the spirit of AdA's XLand to give the "goal" as an observation
+        # but let the agent meta-learn the "rules"... though it's not clear if that
+        # makes xland minigrid easier than it was meant to be...
```

Comment:
From my experience with simple RL^2 baseline, explicit conditioning on true rules and goals makes it a lot easier. While AdA indeed revealed them, they used random masking during training + completely masked it out on evaluation (which is I think right thing to do).